File: Ruby/life.rb
require 'colorize'

WIDTH = 40
HEIGHT = 40
ACTUAL_WIDTH = WIDTH + 2
ACTUAL_HEIGHT = HEIGHT + 2
INITIAL_LIVES = 500
UNIVERSE = Array.new((ACTUAL_HEIGHT) * (ACTUAL_HEIGHT), 0)

# to avoid using too many conditionals when computing the generations, there's a row of 0s all around the universe
# this is to check whether we are at the edge or not, if so we skip to the next cell
def is_edge?(i)
    (i < ACTUAL_WIDTH) || # top row
    (i > (ACTUAL_WIDTH * ACTUAL_HEIGHT) - ACTUAL_WIDTH) || # bottom row
    (i % ACTUAL_WIDTH == 0) || # first column
    ((i + 1) % ACTUAL_WIDTH == 0) # last column
end

def initial_generation
    indexes = []
    UNIVERSE.each_with_index do |x, i|
        if is_edge?(i)
            next
        end
        indexes += [i]
    end
    indexes = indexes.shuffle.take(INITIAL_LIVES)
    indexes.each do |i|
        UNIVERSE[i] = 1
    end
end

def print_universe
    str = ""
    counter = 0
    UNIVERSE.each_with_index do |x, i|
        if is_edge?(i)
            next
        end
        str += "| #{x == 0 ? x : 'X'.green} "
        counter += 1
        if counter % HEIGHT == 0
            str += "| \n"
            puts str
            str = ""
        end
    end
end

def new_generation
    deaths = []
    births = []
    UNIVERSE.each_with_index do |x, i|
        if is_edge?(i)
            next
        end
        living_neighbors = UNIVERSE[i - 1] + # left & right
            UNIVERSE[i + 1] +
            UNIVERSE[i + ACTUAL_WIDTH] + # over & under
            UNIVERSE[i - ACTUAL_WIDTH] +
            UNIVERSE[i + ACTUAL_WIDTH + 1] + # diagonals
            UNIVERSE[i + ACTUAL_WIDTH - 1] +
            UNIVERSE[i - ACTUAL_WIDTH - 1] +
            UNIVERSE[i - ACTUAL_WIDTH + 1]

        if x && (living_neighbors < 2 || living_neighbors > 3)
            deaths += [i]
        elsif !x && (living_neighbors == 2 || living_neighbors == 3)
            births += [i]
        end
    end
    deaths.each do |d|
        UNIVERSE[d] = 0
    end
    births.each do |b|
        UNIVERSE[b] = 1
    end
end

initial_generation
print_universe
50.times do
    system("clear")
    new_generation
    print_universe
    sleep(1)
end

File: Python/life.py
from random import shuffle
from time import sleep
from os import system

WIDTH = 40
HEIGHT = 40
ACTUAL_WIDTH = WIDTH + 2
ACTUAL_HEIGHT = HEIGHT + 2
INITIAL_LIVES = 500
UNIVERSE = [0 for i in range(ACTUAL_HEIGHT * ACTUAL_WIDTH)]


def is_edge(i):
    return (i < ACTUAL_WIDTH) or \
        (i > (ACTUAL_WIDTH * ACTUAL_HEIGHT) - ACTUAL_WIDTH) or \
        (i % ACTUAL_WIDTH == 0) or \
        ((i + 1) % ACTUAL_WIDTH == 0)

def initial_generation():
    indexes = []
    for i in range(len(UNIVERSE)):
        if is_edge(i):
            continue
        indexes.append(i)
    shuffle(indexes)
    for i in indexes[:INITIAL_LIVES]:
        UNIVERSE[i] = 1

def print_universe():
    print_str = ""
    counter = 0
    for i in range(len(UNIVERSE)):
        if is_edge(i):
            continue
        colored_cell = "\033[92mX\033[0m"
        print_str += F"| {'0' if UNIVERSE[i] == 0 else colored_cell} "
        counter += 1
        if counter % HEIGHT == 0:
            print_str += "| \n"
            print(print_str)
            print_str = ""

def new_generation():
    births = []

    for i in range(len(UNIVERSE)):
        if is_edge(i):
            continue
        living_neighbors = UNIVERSE[i - 1] + \
            UNIVERSE[i + 1] + \
            UNIVERSE[i + ACTUAL_WIDTH] + \
            UNIVERSE[i - ACTUAL_WIDTH] + \
            UNIVERSE[i + ACTUAL_WIDTH + 1] + \
            UNIVERSE[i + ACTUAL_WIDTH - 1] + \
            UNIVERSE[i - ACTUAL_WIDTH - 1] + \
            UNIVERSE[i - ACTUAL_WIDTH + 1]
        if living_neighbors == 2 or living_neighbors == 3:
            births.append(i)

    for c in range(0, len(UNIVERSE)):
        if is_edge(c):
            continue
        if c in births:
            UNIVERSE[c] = 1
        else:
            UNIVERSE[c] = 0

if __name__=="__main__":
    initial_generation()
    print_universe()
    for i in range(50):
        system("clear")
        new_generation()
        print_universe()
        sleep(1)